src/Array.mo

```diff
@@ -38,6 +39,64 @@ module {
       };
     };
   };
+
+  /// Sorts the given array according to the `cmp` function.
+  /// This is a _stable_ sort.
+  ///
+  /// ```motoko
+  /// import Array "mo:base/Array";
+  /// import Nat "mo:base/Nat";
+  /// let xs = [4, 2, 6, 1, 5];
+  /// assert(Array.sort(xs, Nat.compare) == [1, 2, 4, 5, 6])
+  /// ```
+  public func sort<A>(xs : [A], cmp : (A, A) -> Order.Order) : [A] {
+    let tmp : [var A] = thaw(xs);
+    sortInPlace(tmp, cmp);
+    freeze(tmp)
```

Comment:
If only we had a linear type system… :-)

Reply:
Yes... having written a bit of Rust these things start jumping out at me as well.

Reply:
Yes... having written a bit of Rust these things start jumping out at me as well.

Reply:
Or a bit of Haskell (`unsafeFreeze` :-))

Reply:
Yeah... Ideally I wouldn't have to initialize `aux` either, but not even Rust would let me do that without unsafe.